pkg/tests/end_to_end_tests/trace_tree_test.go

```diff
@@ -0,0 +1,381 @@
+	const qry = `
+	select span_id, coalesce(parent_span_id, -1), lvl, path::text
+	from trace_tree('3dadb2bf-0035-433e-b74b-9075cc9260e8') 
+	order by span_id`
+
+	type result struct {
+		SpanId       int64
+		ParentSpanId int64
+		Level        int64
+		Path         string
+	}
+
+	t.Run("trace_tree", func(t *testing.T) {
+		expectedResults := []result{
+			{1, -1, 1, "{1}"},
+			{2, 1, 2, "{1,2}"},
+			{3, 1, 2, "{1,3}"},
+			{4, 1, 2, "{1,4}"},
+			{5, 2, 3, "{1,2,5}"},
+			{6, 2, 3, "{1,2,6}"},
+			{7, 2, 3, "{1,2,7}"},
+			{8, 3, 3, "{1,3,8}"},
+			{9, 3, 3, "{1,3,9}"},
+			{10, 3, 3, "{1,3,10}"},
+			{11, 4, 3, "{1,4,11}"},
+			{12, 4, 3, "{1,4,12}"},
+			{13, 6, 4, "{1,2,6,13}"},
+			{14, 8, 4, "{1,3,8,14}"},
+			{15, 12, 4, "{1,4,12,15}"},
+			{16, 13, 5, "{1,2,6,13,16}"},
+			{17, 13, 5, "{1,2,6,13,17}"},
+			{18, 15, 5, "{1,4,12,15,18}"},
+		}
+		rows, err := db.Query(ctx, qry)
+		require.NoError(t, err, "Error running trace_tree: %v", err)
+		defer rows.Close()
+
+		for _, expected := range expectedResults {
+			require.True(t, rows.Next(), "Expected a row for span id %v but got none", expected.SpanId)
+
+			var actual = result{}
+			err = rows.Scan(&actual.SpanId, &actual.ParentSpanId, &actual.Level, &actual.Path)
+			require.NoError(t, err, "Failed to scan row. Expected row for span id %v: %v", expected.SpanId, err)
+
+			require.Equal(t, expected, actual)
+		}
+		require.False(t, rows.Next(), "Found more result rows than expected")
+	})
+}
+
+func testUpstreamSpans(t *testing.T, ctx context.Context, db *pgxpool.Pool) {
+	type result struct {
+		SpanId       int64
+		ParentSpanId int64
+		Dist         int64
+		Path         string
+	}
+
+	eval := func(t *testing.T, qry string, expectedResults []result) {
+		rows, err := db.Query(ctx, qry)
+		require.NoError(t, err, "Error running upstream_spans: %v", err)
+		defer rows.Close()
+
+		for _, expected := range expectedResults {
+			require.True(t, rows.Next(), "Expected a row for span id %v but got none", expected.SpanId)
+
+			var actual = result{}
+			err = rows.Scan(&actual.SpanId, &actual.ParentSpanId, &actual.Dist, &actual.Path)
+			require.NoError(t, err, "Failed to scan row. Expected row for span id %v: %v", expected.SpanId, err)
+
+			require.Equal(t, expected, actual)
+		}
+		require.False(t, rows.Next(), "Found more result rows than expected")
+	}
+
+	t.Run("upstream_spans", func(t *testing.T) {
+		const qry = `
+		select span_id, coalesce(parent_span_id, -1), dist, path::text
+		from upstream_spans('3dadb2bf-0035-433e-b74b-9075cc9260e8', 13)
+		order by span_id`
+		expectedResults := []result{
+			{1, -1, -3, "{1,2,6,13}"},
+			{2, 1, -2, "{2,6,13}"},
+			{6, 2, -1, "{6,13}"},
+			{13, 6, 0, "{13}"},
+		}
+		eval(t, qry, expectedResults)
+	})
+
+	t.Run("upstream_spans_max_dist", func(t *testing.T) {
+		const qry = `
+		select span_id, coalesce(parent_span_id, -1), dist, path::text
+		from upstream_spans('3dadb2bf-0035-433e-b74b-9075cc9260e8', 13, 2)
```

Comment:
In the context of previous discussion on absolute numbers: imho it feels a bit awkward to specify max depth argument as positive number but get negatives in the result.

Reply:
Good point. What do you suggest? Use positive numbers for dist and a separate boolean flag in the results of `span_tree` to indicate direction?

Reply:
will use positive dist across the board and add is_upstream and is_downstream flags